# Kisi Kisi Live Code 2

## Preface

=================================================

Kisi-Kisi Live Code 2

Nama  : Karmenia Ditabaya Lontoh

Batch : HCK-021

Program ini dibuat untuk menampilkan data peminjaman sepeda dengan kondisi-kondisi tertentu, melakukan pengecekan tabel data peminjaman, dan penambahan kolom menggunakan SQL dan Pandas

=================================================

## Answer

### Setup

In [1]:
# Setup koneksi Google Colaboratory dengan BigQuery
from google.colab import auth
from google.cloud import bigquery
auth.authenticate_user()
print('Authenticated')

project_id = "latihan-live-code-2"
client = bigquery.Client(project=project_id)

Authenticated


### SQL

##### 1. Berapa banyak label unique dari bike_type?

In [4]:
# Menggunakan DISTINCT untuk mengambil unique value dari kolom 'bike_type'
# Menggunakan COUNT untuk menghitung total unique value dari kolom 'bike_type'
no1 = client.query("""
SELECT
  COUNT(DISTINCT bike_type) AS total_unique_bike_type
FROM ftds-hacktiv8-project.practice.bike_rent;
""").to_dataframe()

no1

total_unique_bike_type
0                       2

Tipe sepeda memiliki **2 unique value.**

##### 2. Apa saja nama label unique dari tipe sepeda?

In [ ]:
# Menggunakan DISTINCT untuk mengambil nama label unique dari kolom 'bike_type'
no2 = client.query("""
SELECT
  DISTINCT bike_type AS unique_bike_type
FROM ftds-hacktiv8-project.practice.bike_rent;
""").to_dataframe()

no2

unique_bike_type
0          classic
1         electric

Tipe sepeda terdiri dari **classic** dan **electric**

##### 3. Berapa banyak total peminjaman sepeda pada tahun 2016 dan 2017?

In [ ]:
# Menggunakan COUNT(*) untuk menghitung semua row data peminjaman
# Menggunakan EXTRACT YEAR untuk extract tahun dari kolom 'start_time'
no3 = client.query("""
SELECT
  COUNT(*) AS total_rent
FROM ftds-hacktiv8-project.practice.bike_rent
WHERE EXTRACT(YEAR FROM start_time) IN (2016, 2017);
""").to_dataframe()

no3

total_rent
0        7993

Total peminjaman sepeda pada tahun 2016 sampai 2017 adalah **7,993 peminjaman**

##### 4. Dimana station yang paling banyak dikunjungi sebagai titik keberangkatan pada tahun 2016?

In [ ]:
# Menggunakan EXTRACT YEAR untuk extract tahun dari kolom 'start_time'
# Menggunakan COUNT untuk menghitung banyak peminjaman
# Menggunakan ORDER BY untuk sorting mulai dari yang terbesar (descending)
# Menggunakan LIMIT untuk mengambil row pertama yang merupakan data terbesar
no4 = client.query("""
SELECT
  start_station_name
FROM ftds-hacktiv8-project.practice.bike_rent
WHERE EXTRACT(YEAR FROM start_time) = 2016
GROUP BY start_station_name
ORDER BY COUNT(start_station_name) DESC
LIMIT 1;
""").to_dataframe()

no4

start_station_name
0  Bullock Museum @ Congress & MLK

Stasiun **Bullock Museum @ Congress & MLK** merupakan stasiun yang paling sering dijadikan titik keberangkatan oleh peminjam pada tahun 2016

##### 5. Berapa nilai minimum, maximum, dan rata-rata dari durasi perjalanan pada keseluruhan data?

In [ ]:
# Menggunakan MIN untuk mendapatkan nilai minimum durasi perjalanan
# Menggunakan MAX untuk mendapatkan nilai maximum durasi perjalanan
# Menggunakan AVG untuk mendapatkan nilai average durasi perjalanan
no5 = client.query("""
SELECT
  MIN(duration_minutes) AS minimum_durasi,
  MAX(duration_minutes) AS maximum_durasi,
  AVG(duration_minutes) AS rata_rata_durasi
FROM ftds-hacktiv8-project.practice.bike_rent;
""").to_dataframe()

no5

minimum_durasi  maximum_durasi  rata_rata_durasi
0               2            2218           35.0995

Peminjaman sepeda pada tahun 2016 sampai 2018 memiliki waktu peminjaman paling singkat **2 menit** dan paling lama selama **2,218 menit**. Rata-rata peminjam melakukan peminjaman sepeda selama **35.09 menit**

##### 6. Berapa rata-rata durasi perjalanan dari subscriber_type `Explorer` dan `Weekender`?

In [3]:
# Menggunakan AVG untuk mendapatkan nilai average durasi perjalanan
no6 = client.query("""
SELECT
  subscriber_type,
  AVG(duration_minutes) as rata_rata_durasi
FROM ftds-hacktiv8-project.practice.bike_rent
WHERE subscriber_type IN ('Explorer','Weekender')
GROUP BY subscriber_type;
""").to_dataframe()

no6

subscriber_type  rata_rata_durasi
0        Explorer         42.866667
1       Weekender         27.636478

Peminjam dengan tipe `Explorer` untuk peminjaman tahun 2016 sampai 2018 rata-rata melakukan peminjaman selama **42.86 menit** sedangkan peminjam dengan tipe `Weekender` rata-rata melakukan peminjaman selaman **27.63 menit**.

##### 7. Tampilkan entry data yang memiliki `duration_time` diatas 4.25 jam.

In [ ]:
# Subquery untuk mengubah data 'duration_minutes' dari satuan menit ke satuan jam
no7 = client.query("""
SELECT *
FROM (
    SELECT *,
    (duration_minutes / 60) AS duration_time
    FROM ftds-hacktiv8-project.practice.bike_rent
    )
WHERE duration_time > 4.25;
""").to_dataframe()

no7

trip_id subscriber_type bike_id bike_type                start_time  \
0   13255686         Walk Up     539   classic 2016-12-25 11:17:16+00:00   
1   13255691         Walk Up     453   classic 2016-12-25 11:18:17+00:00   
2   14534164         Walk Up     664   classic 2017-04-21 20:46:17+00:00   
3   14534142         Walk Up     275   classic 2017-04-21 20:41:54+00:00   
4   11965934         Walk Up     453   classic 2016-09-17 11:55:23+00:00   
..       ...             ...     ...       ...                       ...   
86  14699898         Walk Up     262   classic 2017-05-06 09:08:08+00:00   
87   9516320         Walk Up     157   classic 2016-04-11 13:19:09+00:00   
88  14465838         Walk Up     871   classic 2017-04-16 12:39:51+00:00   
89  15884138         Walk Up     882   classic 2017-09-07 19:11:26+00:00   
90  14087008         Walk Up     221   classic 2017-03-18 00:14:37+00:00   

                         start_station_name  \
0                                Pease Park   
1                                Pease Park   
2                    Davis at Rainey Street   
3                    Davis at Rainey Street   
4           Bullock Museum @ Congress & MLK   
..                                      ...   
86          State Capitol @ 14th & Colorado   
87                             Lavaca & 6th   
88                 South Congress & Academy   
89  Capital Metro HQ - East 5th at Broadway   
90              East 6th at Robert Martinez   

                           end_station_name  duration_minutes  duration_time  
0                                Pease Park               351       5.850000  
1                                Pease Park               350       5.833333  
2                                Pease Park               783      13.050000  
3                                Pease Park               789      13.150000  
4                              Lavaca & 6th               325       5.416667  
..                                      ...               ...            ...  
86          State Capitol @ 14th & Colorado               350       5.833333  
87          State Capitol @ 14th & Colorado               643      10.716667  
88          State Capitol @ 14th & Colorado               684      11.400000  
89  Capital Metro HQ - East 5th at Broadway               292       4.866667  
90  Capital Metro HQ - East 5th at Broadway               757      12.616667  

[91 rows x 9 columns]

Data di atas merupakan data peminjaman dengan durasi pinjam lebih dari **4.25 jam**

### Pandas

In [ ]:
# Load data frame
df = client.query("""
SELECT *
FROM ftds-hacktiv8-project.practice.bike_rent;
""").to_dataframe()

# Menampilkan data frame
df

trip_id subscriber_type bike_id bike_type                start_time  \
0     18896589        Explorer      93   classic 2018-10-28 14:16:14+00:00   
1     13948351        Local365     315   classic 2017-03-07 12:37:33+00:00   
2     16243803        Local365     877   classic 2017-10-21 09:49:48+00:00   
3     16222446         Walk Up     273   classic 2017-10-18 15:01:52+00:00   
4     16222451         Walk Up     201   classic 2017-10-18 15:02:26+00:00   
...        ...             ...     ...       ...                       ...   
7995  15469480         Walk Up     327   classic 2017-07-22 20:17:46+00:00   
7996  16130650         Walk Up     528   classic 2017-10-07 11:25:36+00:00   
7997  14204138         Walk Up     235   classic 2017-03-27 13:19:02+00:00   
7998  15037809         Walk Up     206   classic 2017-06-05 11:42:40+00:00   
7999   9499076         Walk Up     315   classic 2016-04-09 18:52:09+00:00   

                           start_station_name  \
0                     Dean Keeton & Speedway    
1                                 Main Office   
2           Republic Square @ 5th & Guadalupe   
3             Bullock Museum @ Congress & MLK   
4             Bullock Museum @ Congress & MLK   
...                                       ...   
7995          Bullock Museum @ Congress & MLK   
7996                  Congress & Cesar Chavez   
7997          State Capitol @ 14th & Colorado   
7998          State Capitol @ 14th & Colorado   
7999  Capital Metro HQ - East 5th at Broadway   

                             end_station_name  duration_minutes  
0                              11th & Salina                178  
1                                 Main Office               198  
2                                 Main Office                60  
3                                 Main Office                60  
4                                 Main Office                60  
...                                       ...               ...  
7995          Bullock Museum @ Congress & MLK                55  
7996          Bullock Museum @ Congress & MLK                55  
7997          State Capitol @ 14th & Colorado                55  
7998          State Capitol @ 14th & Colorado                55  
7999  Capital Metro HQ - East 5th at Broadway                55  

[8000 rows x 8 columns]

##### 1. Cek kesesuaian tipe data dan keberadaan missing values dengan `.info()`

In [ ]:
# Menampilkan informasi tabel dengan `.info()`
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   trip_id             8000 non-null   Int64              
 1   subscriber_type     8000 non-null   object             
 2   bike_id             8000 non-null   object             
 3   bike_type           8000 non-null   object             
 4   start_time          8000 non-null   datetime64[us, UTC]
 5   start_station_name  8000 non-null   object             
 6   end_station_name    8000 non-null   object             
 7   duration_minutes    8000 non-null   Int64              
dtypes: Int64(2), datetime64[us, UTC](1), object(5)
memory usage: 515.8+ KB


Berdasarkan informasi tabel, dapat dilihat bahwa:
1. Terdapat **tipe data yang salah**, yaitu tipe data kolom `bike_id` yang seharusnya integer.
2. **Tidak terdapat missing values**, ditandai dengan informasi bahwa seluruh kolom memiliki 8,000 non-null dari total 8,000 entries

##### 2. Buat kolom baru bernama `quarter`.

In [ ]:
# Inisiasi insance untuk define month dari kolom `start_time`
month_num = df['start_time'].dt.month

# Inisiasi list untuk menyimpan data kuartal
kuartal = []

# Loop untuk menentukan kuartal berdasarkan bulan
for month in month_num:
    if month in [1, 2, 3]:
        kuartal.append('Q1')
    elif month in [4, 5, 6]:
        kuartal.append('Q2')
    elif month in [7, 8, 9]:
        kuartal.append('Q3')
    elif month in [10, 11, 12]:
        kuartal.append('Q4')

# Membuat kolom baru `quarter` dengan hasil kuartal
df['quarter'] = kuartal

# Menampilkan data frame yang sudah ditambahkan kolom `quarter`
df

trip_id subscriber_type bike_id bike_type                start_time  \
0     18896589        Explorer      93   classic 2018-10-28 14:16:14+00:00   
1     13948351        Local365     315   classic 2017-03-07 12:37:33+00:00   
2     16243803        Local365     877   classic 2017-10-21 09:49:48+00:00   
3     16222446         Walk Up     273   classic 2017-10-18 15:01:52+00:00   
4     16222451         Walk Up     201   classic 2017-10-18 15:02:26+00:00   
...        ...             ...     ...       ...                       ...   
7995  15469480         Walk Up     327   classic 2017-07-22 20:17:46+00:00   
7996  16130650         Walk Up     528   classic 2017-10-07 11:25:36+00:00   
7997  14204138         Walk Up     235   classic 2017-03-27 13:19:02+00:00   
7998  15037809         Walk Up     206   classic 2017-06-05 11:42:40+00:00   
7999   9499076         Walk Up     315   classic 2016-04-09 18:52:09+00:00   

                           start_station_name  \
0                     Dean Keeton & Speedway    
1                                 Main Office   
2           Republic Square @ 5th & Guadalupe   
3             Bullock Museum @ Congress & MLK   
4             Bullock Museum @ Congress & MLK   
...                                       ...   
7995          Bullock Museum @ Congress & MLK   
7996                  Congress & Cesar Chavez   
7997          State Capitol @ 14th & Colorado   
7998          State Capitol @ 14th & Colorado   
7999  Capital Metro HQ - East 5th at Broadway   

                             end_station_name  duration_minutes quarter  
0                              11th & Salina                178      Q4  
1                                 Main Office               198      Q1  
2                                 Main Office                60      Q4  
3                                 Main Office                60      Q4  
4                                 Main Office                60      Q4  
...                                       ...               ...     ...  
7995          Bullock Museum @ Congress & MLK                55      Q3  
7996          Bullock Museum @ Congress & MLK                55      Q4  
7997          State Capitol @ 14th & Colorado                55      Q1  
7998          State Capitol @ 14th & Colorado                55      Q2  
7999  Capital Metro HQ - East 5th at Broadway                55      Q2  

[8000 rows x 9 columns]

Sudah terdapat kolom baru, yaitu kolom `quarter` yang menjelaskan tentang peminjaman dilakukan di kuarter ke berapa.

##### 3. Tampilkan data perjalanan dari tipe member `Local31` dengan durasi perjalanan diatas 1 jam.


In [ ]:
# Menampilkan data perjalanan dengan filtering `subscriber_type` = Local31 dan `duration_minutes` > 60 menit
df[(df['subscriber_type'].str.contains('Local31')) & (df['duration_minutes'] > 60)]


Empty DataFrame
Columns: [trip_id, subscriber_type, bike_id, bike_type, start_time, start_station_name, end_station_name, duration_minutes, quarter]
Index: []

**Tidak terdapat data** peminjaman oleh member `Local31` yang melakukan peminjaman selama lebih dari 1 jam

##### 4. Berapa rata-rata durasi perjalanan dari tipe pelanggan `Walk Up` dengan tipe sepeda `classic` pada bulan `Maret` dari data yang diambil?


In [ ]:
# Filtering data untuk kondisi `subscriber_type` = Walk Up, `bike_type` = classic, dan `start_time` = 3 (bulan Maret)
no4 = df[ (df['subscriber_type'] == 'Walk Up') &
          (df['bike_type'] == 'classic') &
          (df['start_time'].dt.month == 3)]

# Menampilkan rata-rata (mean) dari `duration_minutes` berdasarkan data yang sudah dilakukan filtering
no4['duration_minutes'].mean()

38.476584022038566

Rata-rata peminjam `Walk Up` yang meminjam sepeda `classic` pada bulan `Maret`, melakukan peminjaman selama **38.47 menit**